File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
class ListNode:
     def __init__(self, val=0, next=None):
         self.val = val
         self.next = next

class Solution:
    def mergeTwoLists2(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:

        dummy = ListNode(-1)
        ans = dummy
        while list1 and list2:
            if list1.val > list2.val:
                ans.next = list2
                list2 = list2.next
            else:
                ans.next = list1
                list1 = list1.next
            ans = ans.next
        if list1:
            ans.next = list1
        else:
            ans.next = list2
        return dummy.next


    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode()
        dummy = head
        while list1 and list2:
            if list1.val < list2.val:
                dummy.next = list1
                list1 = list1.next
            else:
                dummy.next = list2
                list2 = list2.next
            dummy = dummy.next
        if list1:
            dummy.next = list1
        else:
            dummy.next = list2
        return head.next
             
'''
    list1 = [1,2,4], list2 = [1,3,4]

    list1 = 
    list2
'''        

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode dummy;
        ListNode *tmp = &dummy;

        ListNode* l1 = list1;
        ListNode* l2 = list2;

        while(l1 && l2) {
            if(l1->val < l2->val) {
                tmp->next = l1;
                l1 = l1->next;
            } else {
                tmp->next = l2;
                l2 = l2->next;
            }
            tmp = tmp->next;
        }
        while(l1) {
            tmp->next = l1;
            l1=l1->next;
            tmp = tmp->next;
        }
        while(l2) {
            tmp->next = l2;
            l2 = l2->next;
            tmp = tmp->next;
        }

        return dummy.next;
    }
};

/*

dummy

while(l1 && l2) 
    if l1.val < l2.val
        dummy.next = l1
        l1 = l1.next
    else
        dummy.next = l2
        l2 = l2.next

while(l1)
    dummy.next = l1
    l1 = l1.next
while(l2)
    dummy.next = l2
    le = l2.next

*/